File: cli/python/keywords.py
"""
Dictionary-free keyword extractor — Python port of the JS version in
public/lib/mdp-viewer.mjs (KeywordExtractor + detectStructuralTags).

Algorithm: N-gram (CJK 2~8) + ASCII tokens, scored by PMI cohesion x
min(left, right) Shannon entropy. Based on the "無詞典新詞發現" method
(see memory/reference_keyword_pdf.md for the authoritative reference).

Kept structurally identical to the JS implementation so Markdown+ documents
yield the same auto-keywords whether they are viewed in the HTML viewer,
queried from the MCP server, or hit through the HTTP API. If you change
parameters or scoring here, change them in mdp-viewer.mjs too.
"""
from __future__ import annotations

import math
import re

_KEYWORD_SENTENCE_SPLIT_RE = re.compile(
    r"[\s　，。!?！？；：、,.;:()\[\]【】"
    r"「」『』“”‘’《》<>/\\\n\r\t]+"
)
_ASCII_TOKEN_RE = re.compile(r"[A-Za-z][A-Za-z0-9+\-]*")
_ASCII_WORD_RE = re.compile(r"^[A-Za-z][A-Za-z0-9+\-]*$")
_BLOCK_HEADER_LINE_RE = re.compile(r"^\s*-\s+\*\*#[a-z0-9][a-z0-9-]*\*\*.*$", re.MULTILINE)
_HEADING_RE = re.compile(r"^\s*#{1,6}\s+", re.MULTILINE)
_TABLE_ROW_RE = re.compile(r"^\s*\|.*\|\s*$")
_TABLE_SEP_RE = re.compile(r"^\s*\|?\s*:?-+:?\s*(\|\s*:?-+:?\s*)+\|?\s*$")
_MERMAID_BLOCK_RE = re.compile(
    r"^[ \t]*```[ \t]*mermaid\b[ \t]*\r?\n([\s\S]*?)^[ \t]*```",
    re.MULTILINE,
)
_SVG_FENCE_RE = re.compile(r"^[ \t]*```[ \t]*svg\b", re.MULTILINE)
_MERMAID_DIAGRAM_ID_RE = re.compile(r"^([A-Za-z][A-Za-z0-9-]*)")


def _is_cjk(ch: str) -> bool:
    if not ch:
        return False
    c = ord(ch)
    return (
        0x4E00 <= c <= 0x9FFF
        or 0x3400 <= c <= 0x4DBF
        or 0xF900 <= c <= 0xFAFF
    )


def _entropy(counts: dict) -> float:
    total = sum(counts.values())
    if total == 0:
        return 0.0
    ent = 0.0
    for v in counts.values():
        p = v / total
        ent -= p * math.log(p)
    return ent


def clean_text(text: str) -> str:
    """Strip markdown noise that should not contribute to N-gram statistics.
    Mirrors KeywordExtractor.cleanText in mdp-viewer.mjs."""
    s = str(text)
    s = re.sub(r"```[\s\S]*?```", " ", s)
    s = re.sub(r"~~~[\s\S]*?~~~", " ", s)
    s = re.sub(r"`[^`]*`", " ", s)
    s = re.sub(r"!\[[^\]]*\]\([^)]+\)", " ", s)
    s = re.sub(r"\[[^\]]+\]\([^)]+\)", " ", s)
    s = _BLOCK_HEADER_LINE_RE.sub(" ", s)
    s = _HEADING_RE.sub(" ", s)
    s = re.sub(r"[*_>|]+", " ", s)
    return s


class KeywordExtractor:
    """Fits on the whole document, then scores candidates per-block."""

    def __init__(
        self,
        *,
        min_len: int = 2,
        max_len: int = 8,
        min_freq: int = 2,
        min_pmi: float = 1.0,
        min_entropy: float = 0.4,
        min_ascii_len: int = 3,
    ):
        self.min_len = min_len
        self.max_len = max_len
        self.min_freq = min_freq
        self.min_pmi = min_pmi
        self.min_entropy = min_entropy
        self.min_ascii_len = min_ascii_len
        self.ngram_freq: dict[str, int] = {}
        self.left_neighbors: dict[str, dict[str, int]] = {}
        self.right_neighbors: dict[str, dict[str, int]] = {}
        self.total_chars = 0

    def _inc(self, k: str) -> None:
        self.ngram_freq[k] = self.ngram_freq.get(k, 0) + 1

    def _add_neighbor(self, cand: str, left: str, right: str) -> None:
        lm = self.left_neighbors.setdefault(cand, {})
        rm = self.right_neighbors.setdefault(cand, {})
        lm[left] = lm.get(left, 0) + 1
        rm[right] = rm.get(right, 0) + 1

    def fit(self, text: str) -> "KeywordExtractor":
        cleaned = clean_text(text)
        for sent in _KEYWORD_SENTENCE_SPLIT_RE.split(cleaned):
            if len(sent) >= 2:
                self._scan_sentence(sent)
        return self

    def _scan_sentence(self, sent: str) -> None:
        arr = list(sent)
        n = len(arr)
        for i in range(n):
            if _is_cjk(arr[i]):
                self._inc(arr[i])
            self.total_chars += 1
        for i in range(n):
            if not _is_cjk(arr[i]):
                continue
            L = self.min_len
            while L <= self.max_len and i + L <= n:
                ok = True
                for k in range(L):
                    if not _is_cjk(arr[i + k]):
                        ok = False
                        break
                if not ok:
                    break
                cand = "".join(arr[i:i + L])
                self._inc(cand)
                left = arr[i - 1] if i > 0 else "<BOS>"
                right = arr[i + L] if i + L < n else "<EOS>"
                self._add_neighbor(cand, left, right)
                L += 1
        for m in _ASCII_TOKEN_RE.finditer(sent):
            tok = m.group(0)
            if len(tok) < self.min_ascii_len:
                continue
            self._inc(tok)
            start, end = m.start(), m.end()
            left = sent[start - 1] if start > 0 else "<BOS>"
            right = sent[end] if end < len(sent) else "<EOS>"
            self._add_neighbor(tok, left, right)

    def _prob(self, token: str) -> float:
        return self.ngram_freq.get(token, 0) / max(self.total_chars, 1)

    def cohesion(self, word: str) -> float:
        if _ASCII_WORD_RE.match(word):
            return math.log(max(self.ngram_freq.get(word, 1), 1)) + 1
        arr = list(word)
        if len(arr) < 2:
            return 0.0
        p_word = self._prob(word)
        if p_word <= 0:
            return 0.0
        min_pmi = math.inf
        for i in range(1, len(arr)):
            pl = self._prob("".join(arr[:i]))
            pr = self._prob("".join(arr[i:]))
            if pl <= 0 or pr <= 0:
                return 0.0
            pmi = math.log(p_word / (pl * pr))
            if pmi < min_pmi:
                min_pmi = pmi
        return 0.0 if min_pmi == math.inf else min_pmi

    def discover(self, *, top_k: int = 80) -> list[dict]:
        results: list[dict] = []
        for word, freq in self.ngram_freq.items():
            if freq < self.min_freq:
                continue
            is_ascii = bool(_ASCII_WORD_RE.match(word))
            if not is_ascii:
                arr_len = len(word)  # CJK 1 char = 1 code point in BMP/SMP relevant ranges
                if arr_len < self.min_len or arr_len > self.max_len:
                    continue
            else:
                if len(word) < self.min_ascii_len:
                    continue
            coh = self.cohesion(word)
            if coh < self.min_pmi:
                continue
            left_ent = _entropy(self.left_neighbors.get(word, {}))
            right_ent = _entropy(self.right_neighbors.get(word, {}))
            if left_ent < self.min_entropy or right_ent < self.min_entropy:
                continue
            score = freq * coh * min(left_ent, right_ent)
            results.append({"word": word, "freq": freq, "score": score})
        results.sort(key=lambda r: -r["score"])
        kept: list[dict] = []
        for r in results:
            suppress = False
            k = 0
            while k < len(kept):
                o = kept[k]
                if o["word"] == r["word"]:
                    k += 1
                    continue
                if o["word"].find(r["word"]) >= 0 and o["freq"] >= r["freq"] * 0.6:
                    suppress = True
                    break
                if (
                    r["word"].find(o["word"]) >= 0
                    and r["freq"] >= o["freq"] * 0.6
                    and len(r["word"]) > len(o["word"])
                ):
                    kept.pop(k)
                    continue
                k += 1
            if not suppress:
                kept.append(r)
            if len(kept) >= top_k:
                break
        return kept

    def keywords_for_block(
        self,
        block_text: str,
        candidates: list[dict],
        top_n: int = 5,
    ) -> list[str]:
        cleaned = clean_text(block_text)
        hits: list[dict] = []
        for c in candidates:
            w = c["word"]
            cnt = 0
            idx = 0
            while True:
                idx = cleaned.find(w, idx)
                if idx < 0:
                    break
                cnt += 1
                idx += len(w)
            if cnt > 0:
                hits.append({
                    "word": w,
                    "score": c["score"] * math.log(cnt + 1),
                })
        hits.sort(key=lambda h: -h["score"])
        out: list[str] = []
        for h in hits:
            skip = False
            for o in out:
                if o.find(h["word"]) >= 0 or h["word"].find(o) >= 0:
                    skip = True
                    break
            if not skip:
                out.append(h["word"])
            if len(out) >= top_n:
                break
        return out


# --- structural tag detector (port of detectStructuralTags) ---

def _extract_table_column_headers(body_text: str) -> list[str]:
    lines = body_text.split("\n")
    for i in range(len(lines) - 1):
        header_line = lines[i]
        sep_line = lines[i + 1] if i + 1 < len(lines) else ""
        if _TABLE_ROW_RE.match(header_line) and _TABLE_SEP_RE.match(sep_line):
            body = header_line.strip()
            if body.startswith("|"):
                body = body[1:]
            if body.endswith("|"):
                body = body[:-1]
            out: list[str] = []
            for raw in body.split("|"):
                cleaned = raw
                cleaned = re.sub(r"`[^`]*`", "", cleaned)
                cleaned = re.sub(r"\*\*([^*]+)\*\*", r"\1", cleaned)
                cleaned = re.sub(r"\*([^*]+)\*", r"\1", cleaned)
                cleaned = re.sub(r"^[\s—–-]+|[\s—–-]+$", "", cleaned).strip()
                if 1 <= len(cleaned) <= 24 and re.search(r"[A-Za-z一-鿿0-9]", cleaned):
                    out.append(cleaned)
            return out
    return []


def detect_structural_tags(body_text: str, block_type: str) -> list[str]:
    """Pull out non-prose signals — mermaid diagram type, svg fence, table headers.
    Mirrors detectStructuralTags in mdp-viewer.mjs."""
    tags: list[str] = []
    for mm in _MERMAID_BLOCK_RE.finditer(body_text):
        if "mermaid" not in tags:
            tags.append("mermaid")
        for raw_line in mm.group(1).split("\n"):
            line = raw_line.strip()
            if not line:
                continue
            if line.startswith("%%") or line.startswith("---"):
                continue
            id_m = _MERMAID_DIAGRAM_ID_RE.match(line)
            if id_m:
                dt = id_m.group(1)
                if dt != "title" and dt not in tags:
                    tags.append(dt)
            break
    if _SVG_FENCE_RE.search(body_text) and "svg" not in tags:
        tags.append("svg")
    if block_type in ("table", "targets"):
        for h in _extract_table_column_headers(body_text):
            if h not in tags:
                tags.append(h)
    return tags

File: cli/node/keywords.mjs
// Dictionary-free keyword extractor — Node ESM port mirroring
// cli/python/keywords.py and public/lib/mdp-viewer.mjs (KeywordExtractor +
// detectStructuralTags).
//
// All three implementations must yield the same auto-keywords for the same
// source. Algorithm: N-gram (CJK 2-8) + ASCII tokens, scored by PMI cohesion
// x min(left, right) Shannon entropy ("無詞典新詞發現" method).
// If you change parameters or scoring here, change them in the other two too.

const KEYWORD_SENTENCE_SPLIT_RE = /[\s　，。!?！？；：、,.;:()\[\]【】「」『』""''《》<>\/\\\n\r\t]+/;
const TABLE_ROW_RE = /^\s*\|.*\|\s*$/;
const TABLE_SEP_RE = /^\s*\|?\s*:?-+:?\s*(\|\s*:?-+:?\s*)+\|?\s*$/;
const MERMAID_BLOCK_RE = /^[ \t]*```[ \t]*mermaid\b[ \t]*\r?\n([\s\S]*?)^[ \t]*```/gm;
const SVG_FENCE_RE = /^[ \t]*```[ \t]*svg\b/m;
const MERMAID_DIAGRAM_ID_RE = /^([A-Za-z][A-Za-z0-9-]*)/;

function isCJK(ch) {
  if (!ch) return false;
  const c = ch.codePointAt(0);
  return (c >= 0x4e00 && c <= 0x9fff) || (c >= 0x3400 && c <= 0x4dbf) || (c >= 0xf900 && c <= 0xfaff);
}

function entropyOfMap(map) {
  let total = 0;
  for (const v of map.values()) total += v;
  if (total === 0) return 0;
  let ent = 0;
  for (const v of map.values()) {
    const p = v / total;
    ent -= p * Math.log(p);
  }
  return ent;
}

export class KeywordExtractor {
  constructor(opts = {}) {
    this.minLen = opts.minLen ?? 2;
    this.maxLen = opts.maxLen ?? 8;
    this.minFreq = opts.minFreq ?? 2;
    this.minPmi = opts.minPmi ?? 1.0;
    this.minEntropy = opts.minEntropy ?? 0.4;
    this.minAsciiLen = opts.minAsciiLen ?? 3;
    this.ngramFreq = new Map();
    this.leftNeighbors = new Map();
    this.rightNeighbors = new Map();
    this.totalChars = 0;
  }
  _inc(k) { this.ngramFreq.set(k, (this.ngramFreq.get(k) || 0) + 1); }
  _addNeighbor(cand, left, right) {
    let lm = this.leftNeighbors.get(cand);
    if (!lm) { lm = new Map(); this.leftNeighbors.set(cand, lm); }
    let rm = this.rightNeighbors.get(cand);
    if (!rm) { rm = new Map(); this.rightNeighbors.set(cand, rm); }
    lm.set(left, (lm.get(left) || 0) + 1);
    rm.set(right, (rm.get(right) || 0) + 1);
  }
  static cleanText(text) {
    return String(text)
      .replace(/```[\s\S]*?```/g, " ")
      .replace(/~~~[\s\S]*?~~~/g, " ")
      .replace(/`[^`]*`/g, " ")
      .replace(/!\[[^\]]*\]\([^)]+\)/g, " ")
      .replace(/\[[^\]]+\]\([^)]+\)/g, " ")
      .replace(/^\s*-\s+\*\*#[a-z0-9][a-z0-9-]*\*\*.*$/gm, " ")
      .replace(/^\s*#{1,6}\s+/gm, " ")
      .replace(/[*_>|]+/g, " ");
  }
  fit(text) {
    const cleaned = KeywordExtractor.cleanText(text);
    const sentences = cleaned.split(KEYWORD_SENTENCE_SPLIT_RE).filter((s) => s.length >= 2);
    for (const sent of sentences) this._scanSentence(sent);
    return this;
  }
  _scanSentence(sent) {
    const arr = Array.from(sent);
    const n = arr.length;
    for (let i = 0; i < n; i++) {
      if (isCJK(arr[i])) this._inc(arr[i]);
      this.totalChars++;
    }
    for (let i = 0; i < n; i++) {
      if (!isCJK(arr[i])) continue;
      for (let L = this.minLen; L <= this.maxLen && i + L <= n; L++) {
        let ok = true;
        for (let k = 0; k < L; k++) {
          if (!isCJK(arr[i + k])) { ok = false; break; }
        }
        if (!ok) break;
        const cand = arr.slice(i, i + L).join("");
        this._inc(cand);
        const left = i > 0 ? arr[i - 1] : "<BOS>";
        const right = i + L < n ? arr[i + L] : "<EOS>";
        this._addNeighbor(cand, left, right);
      }
    }
    const tokenRe = /[A-Za-z][A-Za-z0-9+\-]*/g;
    let m;
    while ((m = tokenRe.exec(sent)) !== null) {
      const tok = m[0];
      if (tok.length < this.minAsciiLen) continue;
      this._inc(tok);
      const startIdx = m.index;
      const endIdx = m.index + tok.length;
      const left = startIdx > 0 ? sent[startIdx - 1] : "<BOS>";
      const right = endIdx < sent.length ? sent[endIdx] : "<EOS>";
      this._addNeighbor(tok, left, right);
    }
  }
  _prob(token) { return (this.ngramFreq.get(token) || 0) / Math.max(this.totalChars, 1); }
  cohesion(word) {
    if (/^[A-Za-z][A-Za-z0-9+\-]*$/.test(word)) {
      return Math.log(this.ngramFreq.get(word) || 1) + 1;
    }
    const arr = Array.from(word);
    if (arr.length < 2) return 0;
    const pWord = this._prob(word);
    if (pWord <= 0) return 0;
    let minPmi = Infinity;
    for (let i = 1; i < arr.length; i++) {
      const pl = this._prob(arr.slice(0, i).join(""));
      const pr = this._prob(arr.slice(i).join(""));
      if (pl <= 0 || pr <= 0) return 0;
      const pmi = Math.log(pWord / (pl * pr));
      if (pmi < minPmi) minPmi = pmi;
    }
    return minPmi === Infinity ? 0 : minPmi;
  }
  discover({ topK = 80 } = {}) {
    const results = [];
    for (const [word, freq] of this.ngramFreq) {
      if (freq < this.minFreq) continue;
      const isAscii = /^[A-Za-z][A-Za-z0-9+\-]*$/.test(word);
      if (!isAscii) {
        const arr = Array.from(word);
        if (arr.length < this.minLen || arr.length > this.maxLen) continue;
      } else if (word.length < this.minAsciiLen) continue;
      const coh = this.cohesion(word);
      if (coh < this.minPmi) continue;
      const leftEnt = entropyOfMap(this.leftNeighbors.get(word) || new Map());
      const rightEnt = entropyOfMap(this.rightNeighbors.get(word) || new Map());
      if (leftEnt < this.minEntropy || rightEnt < this.minEntropy) continue;
      const score = freq * coh * Math.min(leftEnt, rightEnt);
      results.push({ word, freq, score });
    }
    results.sort((a, b) => b.score - a.score);
    const kept = [];
    for (const r of results) {
      let suppress = false;
      for (let k = 0; k < kept.length; k++) {
        const o = kept[k];
        if (o.word === r.word) continue;
        if (o.word.includes(r.word) && o.freq >= r.freq * 0.6) { suppress = true; break; }
        if (r.word.includes(o.word) && r.freq >= o.freq * 0.6 && r.word.length > o.word.length) {
          kept.splice(k, 1); k--;
        }
      }
      if (!suppress) kept.push(r);
      if (kept.length >= topK) break;
    }
    return kept;
  }
  keywordsForBlock(blockText, candidates, topN = 5) {
    const cleaned = KeywordExtractor.cleanText(blockText);
    const hits = [];
    for (const c of candidates) {
      let cnt = 0;
      let idx = 0;
      while ((idx = cleaned.indexOf(c.word, idx)) !== -1) { cnt++; idx += c.word.length; }
      if (cnt > 0) hits.push({ word: c.word, score: c.score * Math.log(cnt + 1) });
    }
    hits.sort((a, b) => b.score - a.score);
    const out = [];
    for (const h of hits) {
      let skip = false;
      for (const o of out) {
        if (o.includes(h.word) || h.word.includes(o)) { skip = true; break; }
      }
      if (!skip) out.push(h.word);
      if (out.length >= topN) break;
    }
    return out;
  }
}

// --- structural tag detector (port of detectStructuralTags) ---

function extractTableColumnHeaders(bodyText) {
  const lines = bodyText.split(/\r?\n/);
  for (let i = 0; i < lines.length - 1; i++) {
    const headerLine = lines[i];
    const sepLine = lines[i + 1] || "";
    if (TABLE_ROW_RE.test(headerLine) && TABLE_SEP_RE.test(sepLine)) {
      let body = headerLine.trim();
      if (body.startsWith("|")) body = body.slice(1);
      if (body.endsWith("|")) body = body.slice(0, -1);
      const out = [];
      for (const raw of body.split("|")) {
        const cleaned = raw
          .replace(/`[^`]*`/g, "")
          .replace(/\*\*([^*]+)\*\*/g, "$1")
          .replace(/\*([^*]+)\*/g, "$1")
          .replace(/^[\s—–-]+|[\s—–-]+$/g, "")
          .trim();
        if (cleaned.length >= 1 && cleaned.length <= 24 && /[A-Za-z一-鿿0-9]/.test(cleaned)) {
          out.push(cleaned);
        }
      }
      return out;
    }
  }
  return [];
}

export function detectStructuralTags(bodyText, blockType) {
  const tags = [];
  let mm;
  // Reset state — global regex shared across calls.
  MERMAID_BLOCK_RE.lastIndex = 0;
  while ((mm = MERMAID_BLOCK_RE.exec(bodyText)) !== null) {
    if (!tags.includes("mermaid")) tags.push("mermaid");
    for (const rawLine of mm[1].split(/\r?\n/)) {
      const line = rawLine.trim();
      if (!line) continue;
      if (line.startsWith("%%") || line.startsWith("---")) continue;
      const idM = line.match(MERMAID_DIAGRAM_ID_RE);
      if (idM) {
        const dt = idM[1];
        if (dt !== "title" && !tags.includes(dt)) tags.push(dt);
      }
      break;
    }
  }
  if (SVG_FENCE_RE.test(bodyText) && !tags.includes("svg")) tags.push("svg");
  if (blockType === "table" || blockType === "targets") {
    for (const h of extractTableColumnHeaders(bodyText)) {
      if (!tags.includes(h)) tags.push(h);
    }
  }
  return tags;
}
